composer/algorithms/alibi/attention_surgery_functions/utils.py

```diff
@@ -0,0 +1,184 @@
+                    # Note that this function can (and often should for ALiBi) depend on `max_sequence_length`
+
+                    # YOUR CODE HERE
+
+                    return module
+                return convert_attention
+
+    In the above example, by decorating `build_gpt2_attention_converter` (which is an instance of a `SurgeryFunctionBuilder`
+    function) with `@register_surgery_function_builder(GPT2Attention)`, the ALiBi algorithm will now apply model surgery to any
+    instances of `GPT2Attention` within the model, and will apply surgery on those instances using the `convert_attention` function
+    returned by `build_gpt2_attention_converter`.
+
+    Note that `convert_attention` follows the specific signature of a `ReplacementFunction`, which has specific arguments (for additional
+    details, see `composer/utils/module_surgery.py`). However, often the model surgery functions used by ALiBi should depend on
+    a `max_sequence_length` argument. Since this argument is not used by a `ReplacementFunction`, we make use of these
+    `SurgeryFunctionBuilder` functions so that the exact functions used in model surgery can properly refer to the `max_sequence_length`
+    argument, which is not provided until runtime (see `../alibi.py`).
+    """
+
+    def _register_module(module: Type[torch.nn.Module], func: Callable) -> None:
+        assert issubclass(module, torch.nn.Module)
+        if module in replacement_policy_mapping_builder:
+            raise ValueError(f'Module {module.__name__} already has a registered SurgeryFunctionBuilder.')
+        replacement_policy_mapping_builder[module] = func
+        return
+
+    def wrapper(func: SurgeryFunctionBuilder) -> SurgeryFunctionBuilder:
+        for module in modules:
+            _register_module(module, func)
+        return func
+
+    return wrapper
+
+
+def zero_and_freeze_expand_position_embeddings(
+    module: torch.nn.Module,
+    max_sequence_length: int,
+    position_embedding_attribute: str,
+) -> None:
+    """Replaces weights with zero tensor and prevents them from being learned further.
+
+    This is intended to be used specifically for "removing" positional embeddings.
+    """
+    try:
+        pos_embedding_module = attrgetter(position_embedding_attribute)(module)
+        old_weight = getattr(pos_embedding_module, 'weight')
+        if not isinstance(old_weight, torch.nn.Parameter):
+            raise TypeError(f'Module {module._get_name()}, position embedding {position_embedding_attribute}, '
+                            f"'weight' attribute must be of type torch.nn.Module")
+        new_weight = torch.nn.Parameter(
+            torch.zeros((max_sequence_length, old_weight.shape[1]),
+                        dtype=old_weight.dtype,
+                        layout=old_weight.layout,
+                        device=old_weight.device))
+        new_weight.requires_grad = False
+        setattr(pos_embedding_module, 'weight', new_weight)
```

Comment:
Do we need to overwrite this with an empty tensor? Can we just delete the attribute?

Reply:
I think that would cause problems because the attribute is referenced in the `forward` method of the parent module. We could replace the `forward` method so that this attribute isn't referenced, but that would add boilerplate.

Reply:
wait, does the parent forward() get called? I'm probably just not understanding the code fully.

Reply:
The parent `forward` gets called when the full model is executed. Maybe one point of confusion is that, in this implementation, none of the modules are swapped out with new modules (which is I think the typical way surgery works). Instead, the modules are modified. 
Alternatively, I could have misunderstood your original comment and created additional confusion with my response.

Reply:
Possible point of confusion: is `parent` referring to the owning module or a parent class?

I was understanding your statement as referring to the latter. Since:

1. for every model we modify, the `forward()` gets completely replaced with the `forward()` function in alibi, and
2. This function doesn't reference `super().forward()`,

I expect no forward() will ever try to access the parameter in question.

But if the owning module accesses this parameter within its child, then that makes total sense.

Reply:
I think the intended interpretation is that "the owning module accesses this parameter within its child." ... I think.

Note, though, that in the course of alibi surgery on a Bert model, we do surgery on the `BertEmbedding` module, but we do not replace the `forward` function when modifying that module type. Instead, we just zero/freeze the weights of its `position_embedding` attribute. `BertEmbedding.forward()` (which references its `self.position_embedding`) remains intact. (see `composer/algorithms/alibi/attention_surgery_functions/_bert.py`)